testing/citest/tests/gcs_pubsub_gae_test.py

```diff
@@ -0,0 +1,453 @@
+# --test_subscription_name <configured subscription name>
+# --test_gcs_bucket <bucket notifying pub/sub topic>
+# --spinnaker_appengine_account
+# --test_app
+# --test_stack
+#
+# pylint: disable=bad-continuation
+# pylint: disable=invalid-name
+# pylint: disable=missing-docstring
+
+# Standard python modules.
+import json
+import logging
+import os
+import subprocess
+import shutil
+import sys
+import tempfile
+import time
+
+# citest modules.
+import citest.base
+import citest.gcp_testing as gcp
+import citest.json_contract as jc
+import citest.json_predicate as jp
+import citest.service_testing as st
+
+# Spinnaker modules.
+import spinnaker_testing as sk
+import spinnaker_testing.gate as gate
+import spinnaker_testing.frigga as frigga
+
+ov_factory = jc.ObservationPredicateFactory()
+
+class GcsPubsubGaeTestScenario(sk.SpinnakerTestScenario):
+  """
+  Scenario for testing GAE deploys of GCS artifacts via pub/sub triggers.
+  """
+  @classmethod
+  def new_agent(cls, bindings):
+    return gate.new_agent(bindings)
+
+  @classmethod
+  def initArgumentParser(cls, parser, defaults=None):
+    """Initialize command line argument parser.
+
+    Args:
+    parser: argparse.ArgumentParser
+    """
+    super(GcsPubsubGaeTestScenario, cls).initArgumentParser(
+      parser, defaults=defaults)
+
+    defaults = defaults or {}
+    parser.add_argument(
+      '--app_directory_root', default=None,
+      help='Path from the root of source code repository to the application directory.')
+    parser.add_argument(
+      '--branch', default='master',
+      help='Git branch to be used when deploying from source code repository.')
+    parser.add_argument(
+      '--git_repo_url', default=None,
+      help='URL of a git source code repository used by Spinnaker to deploy to App Engine.')
+    parser.add_argument(
+      '--test_gcs_bucket', default=None,
+      help='GCS bucket to upload GAE app source code to.')
+    parser.add_argument(
+      '--test_storage_account_name', default=None,
+      help='Storage account when testing GCS buckets.'
+      ' If not specified, use the application default credentials.')
+    parser.add_argument(
+      '--test_subscription_name', default=None,
+      help='Google pub/sub subscription name configured in Echo.')
+
+  def __init__(self, bindings, agent=None):
+    super(GcsPubsubGaeTestScenario, self).__init__(bindings, agent)
+    self.logger = logging.getLogger(__name__)
+
+    bindings = self.bindings
+
+    if not bindings['GIT_REPO_URL']:
+      raise ValueError('Must supply value for --git_repo_url')
+
+    if not bindings['APP_DIRECTORY_ROOT']:
+      raise ValueError('Must supply value for --app_directory_root')
+
+    # We'll call out the app name because it is widely used
+    # because it scopes the context of our activities.
+    self.TEST_APP = bindings['TEST_APP']
+    self.TEST_STACK = bindings['TEST_STACK']
+    self.EXPECTED_ARTIFACT_ID = 'deployable-gae-app-artifact'
```

Comment:
Is this arbitrary? If so you might want to tie it to this app so it is easier to trace, can have mulitple tests running, and arent affected by leaks in failed or broken tests.

Reply:
It's arbitrary, but baked into the pipeline only which is scoped to the app. I can parameterize it if you want.